pkg/source/firefly/source.go

```diff
@@ -0,0 +1,156 @@
+		config: config,
+	}, nil
+}
+
+// ExportDevice implements the source.Source interface.
+func (s Source) ExportDevice(devEUI string) (*ttnpb.EndDevice, error) {
+	ffdev, err := devices.GetDevice(devEUI)
+	if err != nil {
+		return nil, err
+	}
+	v3dev := &ttnpb.EndDevice{
+		Name:        ffdev.Name,
+		Description: ffdev.Description,
+		// Formatters:      &ttnpb.MessagePayloadFormatters{},
+		FrequencyPlanId: s.config.frequencyPlanID,
+		Ids: &ttnpb.EndDeviceIdentifiers{
+			ApplicationIds: &ttnpb.ApplicationIdentifiers{ApplicationId: "firefly-" + s.config.appID},
+			DeviceId:       "eui-" + strings.ToLower(ffdev.EUI),
+		},
+		LorawanVersion:    ttnpb.MACVersion_MAC_V1_0_2,
+		LorawanPhyVersion: ttnpb.PHYVersion_RP001_V1_0_2_REV_B,
+		MacSettings: &ttnpb.MACSettings{
+			DesiredAdrAckLimitExponent: &ttnpb.ADRAckLimitExponentValue{Value: ttnpb.ADRAckLimitExponent(ffdev.AdrLimit)},
+			Rx2DataRateIndex:           &ttnpb.DataRateIndexValue{Value: ttnpb.DataRateIndex(ffdev.Rx2DataRate)},
+			StatusCountPeriodicity:     &pbtypes.UInt32Value{Value: 0},
+			StatusTimePeriodicity:      pbtypes.DurationProto(0),
+		},
+		SupportsClassC: ffdev.ClassC,
+		SupportsJoin:   ffdev.ApplicationKey != "",
+	}
+
+	if ffdev.Location != nil {
+		v3dev.Locations = map[string]*ttnpb.Location{
+			"user": {
+				Latitude:  ffdev.Location.Lattitude,
+				Longitude: ffdev.Location.Longitude,
+				Source:    ttnpb.LocationSource_SOURCE_REGISTRY,
+			},
+		}
+		logger.Debugw("Set location", "location", v3dev.Locations)
+	}
+	v3dev.Ids.DevEui, err = unmarshalTextToBytes(&types.EUI64{}, ffdev.EUI)
+	if err != nil {
+		return nil, err
+	}
+	v3dev.Ids.JoinEui, err = unmarshalTextToBytes(&types.EUI64{}, s.config.joinEUI)
+	if err != nil {
+		return nil, err
+	}
+	if v3dev.SupportsJoin {
+		v3dev.RootKeys = &ttnpb.RootKeys{AppKey: &ttnpb.KeyEnvelope{}}
+		v3dev.RootKeys.AppKey.Key, err = unmarshalTextToBytes(&types.AES128Key{}, ffdev.ApplicationKey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	hasSession := ffdev.Address != "" && ffdev.NetworkSessionKey != "" && ffdev.ApplicationSessionKey != ""
+	if hasSession || !v3dev.SupportsJoin {
+		v3dev.Session = &ttnpb.Session{Keys: &ttnpb.SessionKeys{AppSKey: &ttnpb.KeyEnvelope{}, NwkSEncKey: &ttnpb.KeyEnvelope{}}}
+		v3dev.Session.DevAddr, err = unmarshalTextToBytes(&types.DevAddr{}, ffdev.Address)
+		if err != nil {
+			return nil, err
+		}
+		v3dev.Session.Keys.AppSKey.Key, err = unmarshalTextToBytes(&types.AES128Key{}, ffdev.ApplicationSessionKey)
+		if err != nil {
+			return nil, err
+		}
+		v3dev.Session.Keys.NwkSEncKey.Key, err = unmarshalTextToBytes(&types.AES128Key{}, ffdev.NetworkSessionKey)
+		if err != nil {
+			return nil, err
+		}
+		v3dev.Session.LastAFCntDown = uint32(ffdev.FrameCounter)
+		v3dev.Session.LastNFCntDown = uint32(ffdev.FrameCounter)
+		packet, err := devices.GetLastPacket()
+		if err != nil {
+			return nil, err
+		}
+		v3dev.Session.LastFCntUp = uint32(packet.FCnt)
+	}
+
+	if !s.config.dryRun {
```

Comment:
I am inclined to instead have a `--clear-keys` instead. In retrospective, lots of people broke their end devices with `ttnv2` by not knowing that outside of dry runs, the keys are cleared.

Reply:
Making it safe by default seems a good idea to me. `--clear-keys` seems to be related to the implementation, maybe something describing what's happening from a higher point of view like `--disable-devices` _(probably some better options out there)_ could work as a universal solution in the future? Unless that's not the intent.

Reply:
Disable implies that it can be rolled back, which is not really the case when you delete the keys and forget to save the output (this happened _many_ times). At least `--clear-keys` is equivalent to `--i-really-know-what-i-am-doing`.